README.md
# Mapcrafter [![Build Status](https://travis-ci.org/mapcrafter/mapcrafter.svg?branch=master)](https://travis-ci.org/mapcrafter/mapcrafter)

by Moritz Hilscher *(modified by Michael Lavaire)*

### <img src="https://logos-download.com/wp-content/uploads/2016/09/Docker_logo.png" width=48 alt="docker">  https://hub.docker.com/r/miclav/mapcrafter

Mapcrafter is a high performance Minecraft map renderer written in C++. It renders
Minecraft worlds to a bunch of tiles (images) which are viewable as a map in any browser
using Leaflet.js.

<p align="center">
  <img src="https://i.imgur.com/WyKXrgO.png" alt="Example map">
</p>

<!-- TOC -->

- [Features](#features)
- [Requirements](#requirements)
- [Example maps](#example-maps)
- [Help](#help)
  - [Documentation](#documentation)
  - [Reporting bugs and feature requests](#reporting-bugs-and-feature-requests)
  - [Community help](#community-help)
- [Acknowledgements](#acknowledgements)
- [License](#license)
- [Version history](#version-history)

<!-- /TOC -->

## Supported Minecraft versions

The current version supports Minecraft versions 1.18.x. Newer versions may partially work.
For Minecraft 1.17.x (or older) please use previous version of Mapcrafter.

## Features

- **Web output:** Render your Minecraft worlds to maps viewable in any web browser!
- **Different render views:** Choose between different perspectives to render your world
  from! A 2D topdown and a 3D isometric render view are available!
- **Different rotations:** Choose from four different rotations to render your worlds from!
- **Different render modes:** Choose between different render modes like day, night and cave
  for your maps!
- **Different overlays:** Show additional information on your map! For example: Where can
  slimes spawn? Where can monsters spawn at night?
- **Configuration files:** Highly-customizable which worlds are rendered with which render
  view and other render parameters!
- **Markers:** Automatically generated markers from your Minecraft world data!
- **Other stuff:** Biome colors, incremental rendering, multithreading

## Requirements

- A Linux-based or Unix-like operating system like Mac OS
- A decent C++ compiler (preferable `gcc` >= 4.4, or `clang`), CMake and make if you want
  to build Mapcrafter from source
- Some libraries:
  - `libpng`
  - `libjpeg` (but you may use libjpeg-turbo as drop in replacement)
  - `libboost-iostreams`
  - `libboost-system`
  - `libboost-filesystem` (>= 1.42)
  - `libboost-program-options`
  - `libboost-test` (if you want to use the tests)
- For your Minecraft worlds:
  - Anvil world format
  - Minecraft resource packs

## Example maps

There are a few example maps of the renderer on the
[GitHub Wiki](https://github.com/mapcrafter/mapcrafter/wiki/Example-maps).
Please feel free to add your own maps to this list.

## Help

### Documentation

You can read the current documentation on [docs.mapcrafter.org](http://docs.mapcrafter.org)
and you can also [download other builds](https://readthedocs.org/projects/mapcrafter/downloads/)
from there. The documentation source is in the `docs/` directory and you can build it
yourself with Sphinx.

### Reporting bugs and feature requests

If you find bugs or problems when using Mapcrafter or if you have ideas for new
features, then please feel free to add an issue to the [GitHub issue
tracker](https://github.com/mapcrafter/mapcrafter/issues).

### Community help

[Join us on Discord](https://discord.gg/QNe8jXT), where you can meet other Mapcrafter users
and ask for help. Please be respectful with everyone and have patience with people who are
offering their time to help!

## Acknowledgements

Thanks to pigmap and Minecraft Overviewer, whose documentations and source code
were very helpful. I also used the alpha blending code of pigmap and some maps
stuff of the template from Minecraft Overviewer.

## License

Mapcrafter is free software and available under the GPL license.  You can
access the latest source code of Mapcrafter on GitHub:
http://github.com/mapcrafter/mapcrafter

## Version history

See the [CHANGELOG](./CHANGELOG.md) for details.


CHANGELOG.md
# Mapcrafter version history

**v.3.1.328**
  * 1.19-pre1 blocks included
  * Blocks updated with a lot of texture mapping issues resolved in a Blockcrafter update

**v.3.1.323**
* Support for 1.18.x
  * Dockerfile build fixed
  * Entity cache file not saved in world folder anymore. World folder can now be mounted as read only.
  * Breaking change, doesn't support the old chunk format anymore (sorry, :( no time to develop a dynamic convertion of old Minecraft file format)
  * Chunks height supoprted from -64 to 319 included.
  * Support new biome palette
  * Biomes now identified by their name, no more by ID
* Add support for alternate block rotation, improving overall visual, breaking monotonous flat surfaces
* Block atlas, now palettized to minimize memory footprint and disk space
* Performance improvements:
  * Region header IO operations reduced from 4098 to 4 calls.
  * Rendering reworked by separating block render and water layer
  * World rotation now supported on the rendering layer, no more world data rotation
* Bug fixes:
  * Side view of waterloged block wasn't rendering side water
* Still to be worked on:
  * lighting on edges funky
  * few water rendering edge cases
  * Blockcrafter:
    * Textures misplaced on some blocks (ex: cauldron, glass, ...)
    * face sorting in blockcrafter to minimize Z-fighting
    * Missing block types:
      * Banners (doable?)
      * end gate
      * standard heads (skeleton, player, wither skeleton, creeper, dragon)
      * Candle + cake, faces not properly placed
      * Lit candles don't have flames
    * UV's rotation not right in some circonstances

**v.3.0**

* Add support for Minecraft versions 1.13 through 1.16 (#279)

**v.2.4**

* Add support for Minecraft 1.11 and 1.12 blocks (thanks to Bjarno #256)

**v.2.3.1**

* Fixed wrong Leaflet coordinate conversion in topdown and isometric render view when
    using different tiles sizes and fixed wrong chunk <-> tile mapping which caused the
    issue in isometric render view (#198)
  * **Unfortunately you have to force-re-render your isometric maps as the isometric
    chunk <-> tile mapping has slightly changed.**

**v.2.3**

* Added support for Minecraft 1.10 blocks (#226)
* Added new bootstrap web UI (thanks to jamestaylr #205)
* Added `postfix` option for signs to also use a postfix instead of just a prefix (#223)
* Added `--force-render-all` to force-render all maps (#212)
* Fixed map selection on mobile devices (thanks to joshproehl #220 #123)
* Fixed a bug in the bilinear image resizing algorithm that caused dark sides of
    grass and other transparent blocks

**v.2.2.1**

* Updated Minecraft version file to 1.9

**v.2.2**

* Added support for 1.9 blocks (and hoppers)
* Fixed bug which caused water not being rendered (#178)
* Fixed circular crop overflowing bug (#179)
* Fixed a config validation bug (#185)

**v.2.1**

* Added support for corner stairs (thanks for some Minecraft Overviewer code!)
* Added signs to isometric and topdown render view
* Fixed a lighting bug that appeared using some Debian packages
* Fixed wrong slime overlay
* Fixed rotation of pumpkins in isometric render view

**v.2.0**

* Huge internal code refactoring™
* Support for multiple rendering engines: A 2D topdown and the usual 3D isometric render
  view are available
* Support for indexed png images, reducing size of output images drastically
* New overlay render modes (slime chunks, show where monsters can spawn at day/night)
* Configurable tile sizes (to improve performance for small texture sizes)
* Configurable water opacity/transparency
* Restructured internal render management components to allow integration of Mapcrafter
  rendering core into different applications
* Further integration of Mapcrafter into a graphical user interface
  ([mapcrafter-gui](https://github.com/mapcrafter/mapcrafter-gui))
* **There is a new repository for Debian/Ubuntu packages. Please have a look at the
    documentation and update your apt source listings!**

**v.1.5.4**

* Added a detection in the CMake files which compiler C++11 flag to use (for newer
  compilers which do not support the older ``-std=c++0x`` flag)
* Fixed bug with the marker tool (new json sign text format was not parsed properly)
* Fixed bug in texture extraction script (convert command was not properly detected)
* Fixed bug with some textures (#122) by making Mapcrafter able to read indexed PNG images
  and PNG images with different other color formats

**v.1.5.3**

* Better handling of corrupt world files (for now, bigger refactoring of that probably later)
* Added a fix using imagemagick in texture extraction script to prevent libpng warnings
* Fixed bug with boost program options which made it unable to use paths
  containing spaces for the ``--config`` command line option
* Fixed wrong error messages when loading configuration files (Mapcrafter tried
  to read not existing files and even directories if specified as configuration file)

**v.1.5.2**

* Fixed the marker tool to work with the new internal Minecraft 1.8 sign format
* Fixed bug with global logging configuration file installed and searched in wrong directory
* Fixed bug causing the Leaflet map being stuck
* Fixed some small issues with new 1.8 blocks

**v.1.5.1**

* Added ability to configure logging facility with configuration files
* Added colored terminal output for warning/error log messages
* Improved cave rendermode with a new high contrast block coloring (thanks to dtfinch, #100),
  old behavior is still available with a configuration option for compatibility reasons
* Added support for Minecraft 1.8
  * **Since the structure of the textures has changed a bit (especially chest textures),
    you should extract the texture files from a 1.8 Minecraft Jar file to a new, clean
    texture directory.**

**v.1.5**

* Started internal refactoring to improve integration with possible Mapcrafter GUI
  or other programs using Mapcrafter's functionality

src/mapcraftercore/mc/chunk.cpp
/*
 * Copyright 2012-2016 Moritz Hilscher
 *
 * This file is part of Mapcrafter.
 *
 * Mapcrafter is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Mapcrafter is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Mapcrafter.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "chunk.h"
#include "blockstate.h"
#include "../renderer/biomes.h"
#include "../renderer/blockimages.h"

#include <cmath>
#include <iostream>
#include <boost/range.hpp>

namespace mapcrafter {
namespace mc {

namespace {

void readPackedShorts_v116(const std::vector<int64_t>& data, uint16_t* palette, uint16_t* palette_end) {
	uint32_t palette_size = palette_end - palette;
	uint32_t shorts_per_long = (palette_size + data.size() - 1) / data.size();
	uint32_t bits_per_value = 64 / shorts_per_long;
	std::fill(palette, &palette[palette_size], 0);
	uint16_t mask = (1 << bits_per_value) - 1;

	for (uint32_t i=0; i<shorts_per_long; i++) {
		uint32_t j = 0;
		for( uint32_t k=i; k<palette_size; k+=shorts_per_long) {
			assert(j < data.size());
			palette[k] = (uint16_t)(data[j] >> (bits_per_value * i)) & mask;
			j++;
		}
		assert(j <= data.size());
	}
}

} // namespace

uint16_t Chunk::nop_id = 0;

Chunk::Chunk()
	: chunkpos(42, 42) {
	clear();
}

Chunk::~Chunk() {
}

void Chunk::setWorldCrop(const WorldCrop& world_crop) {
	this->world_crop = world_crop;
}

int Chunk::positionToKey(int x, int z, int y) const {
	return y + 256 * (x + 16 * z);
}

bool Chunk::readNBT(mc::BlockStateRegistry& block_registry, const char* data, size_t len,
		nbt::Compression compression) {
	clear();

	// In case it wasn't set before
	if (nop_id == 0) {
		// The no operation block is the air block
		nop_id = block_registry.getBlockID(mc::BlockState("minecraft:air"));
	}

	nbt::NBTFile nbt;
	nbt.readNBT(data, len, compression);

	// Make sure we know which data format this chunk is built of
	if (!nbt.hasTag<nbt::TagInt>("DataVersion")) {
		LOG(ERROR) << "Chunk error: No version tag found!";
		return false;
	}

	int data_version = nbt.findTag<nbt::TagInt>("DataVersion").payload;
	if (data_version < 2860){
		LOG(ERROR) << "Chunk error: Unsupported chunk version, please upgrade.";
		return false;
	}

	// then find x/z pos of the chunk
	if (!nbt.hasTag<nbt::TagInt>("xPos") || !nbt.hasTag<nbt::TagInt>("yPos") || !nbt.hasTag<nbt::TagInt>("zPos")) {
		LOG(ERROR) << "Corrupt chunk: No x/z position found!";
		return false;
	}

	chunkpos = ChunkPos(nbt.findTag<nbt::TagInt>("xPos").payload,
	                             nbt.findTag<nbt::TagInt>("zPos").payload);
	int chunk_lowest = nbt.findTag<nbt::TagInt>("yPos").payload;

	// now we have the original chunk position:
	// check whether this chunk is completely contained within the cropped world
	chunk_completely_contained = world_crop.isChunkCompletelyContained(chunkpos);

	if (nbt.hasTag<nbt::TagString>("Status")) {
		const nbt::TagString& tag = nbt.findTag<nbt::TagString>("Status");
		// completely generated chunks in fresh 1.13 worlds usually have status 'fullchunk' or 'postprocessed'
		// however, chunks of converted <1.13 worlds don't use these, but the state 'mobs_spawned'
		if (!(tag.payload == "fullchunk" || tag.payload == "full" || tag.payload == "postprocessed" || tag.payload == "mobs_spawned")) {
			return true;
		}
	}

	// find sections list
	// ignore it if section list does not exist, can happen sometimes with the empty
	// chunks of the end
	if (!nbt.hasList<nbt::TagCompound>("sections"))
		return true;

	const nbt::TagList& sections_tag = nbt.findTag<nbt::TagList>("sections");
	if (sections_tag.tag_type != nbt::TagCompound::TAG_TYPE)
		return true;

	// go through all sections
	for (auto it = sections_tag.payload.begin(); it != sections_tag.payload.end(); ++it) {
		const nbt::TagCompound& section_tag = (*it)->cast<nbt::TagCompound>();

		// make sure section is valid
		if (!section_tag.hasTag<nbt::TagByte>("Y")
			|| !section_tag.hasTag<nbt::TagCompound>("block_states")
			|| !section_tag.hasTag<nbt::TagCompound>("biomes"))
			continue;

		// Check the Y
		const nbt::TagByte& y = section_tag.findTag<nbt::TagByte>("Y");
		if (y.payload < chunk_lowest || y.payload >= chunk_lowest+Y_CHUNKS_PER_REGION_FILE )
			continue;

		const nbt::TagCompound& blockstates = section_tag.findTag<nbt::TagCompound>("block_states");
		if (!blockstates.hasTag<nbt::TagList>("palette"))
			continue;

		const nbt::TagCompound& biomes = section_tag.findTag<nbt::TagCompound>("biomes");
		if (!biomes.hasTag<nbt::TagList>("palette"))
			continue;

		// create a ChunkSection-object
		ChunkSection section;
		section.y = y.payload;

		/**
		 * Get the block states palette
		 */

		// Depalettize block_states palette
		const nbt::TagList& palettebs = blockstates.findTag<nbt::TagList>("palette");
		std::vector<mc::BlockState> palette_blockstates(palettebs.payload.size());
		std::vector<uint16_t> palette_blockstates_idx(palettebs.payload.size());
		int i = 0;
		for (auto pbsit = palettebs.payload.begin(); pbsit != palettebs.payload.end(); ++pbsit, ++i) {
			nbt::TagCompound& entry = (*pbsit)->cast<nbt::TagCompound>();
			const nbt::TagString& name = entry.findTag<nbt::TagString>("Name");

			mc::BlockState block(name.payload);
			if (entry.hasTag<nbt::TagCompound>("Properties")) {
				const nbt::TagCompound& properties = entry.findTag<nbt::TagCompound>("Properties");
				for (auto it3 = properties.payload.begin(); it3 != properties.payload.end(); ++it3) {
					std::string key = it3->first;
					std::string value = it3->second->cast<nbt::TagString>().payload;
					if (block_registry.isKnownProperty(block.getName(), key)) {
						block.setProperty(key, value);
					}
				}
			}
			palette_blockstates[i] = block;
			palette_blockstates_idx[i] = block_registry.getBlockID(block);
		}

		/**
		 * Get the block states data
		 */
		if (palettebs.payload.size()>1) {
			const nbt::TagLongArray& databs = blockstates.findTag<nbt::TagLongArray>("data");
			readPackedShorts_v116(databs.payload, section.block_ids, &section.block_ids[boost::size(section.block_ids)]);

			bool ok = true;
			for (size_t i = 0; i < 16*16*16; i++) {
				if (section.block_ids[i] >= palette_blockstates.size()) {
					int bits_per_entry = databs.payload.size() * 64 / (16*16*16);
					LOG(ERROR) << "Incorrectly parsed palette ID " << section.block_ids[i]
						<< " at index " << i << " (max is " << palette_blockstates.size()-1
						<< " with " << bits_per_entry << " bits per entry)";
					ok = false;
					break;
				}
				section.block_ids[i] = palette_blockstates_idx[section.block_ids[i]];
			}
			if (!ok) {
				continue;
			}
		} else if (palettebs.payload.size()==1) {
			// Check if air is the only block in this section, if so, ignore it completly, it will speed up the rest
			// of the rendering as we won't have to verify every single block in this section.
			if (palette_blockstates_idx[0] == nop_id) continue;
			// Only 1 in palette: There's only block in this chunk
			std::fill(section.block_ids, section.block_ids+boost::size(section.block_ids), palette_blockstates_idx[0]);
		} else {
			// No palette, this shouldn't happen, anyway let's use the default one
			std::fill(section.block_ids, section.block_ids+boost::size(section.block_ids), 0);
		}


		/**
		 * Get the biome data
		 */
		const nbt::TagList& paletteb = biomes.findTag<nbt::TagList>("palette");
		if (paletteb.payload.size()>1) {
			// More than one biome: there must be data and palette size > 1
			if (!biomes.hasTag<nbt::TagLongArray>("data"))
				continue;
			const nbt::TagLongArray& databiomes = biomes.findTag<nbt::TagLongArray>("data");
			readPackedShorts_v116(databiomes.payload, section.biomes, &section.biomes[boost::size(section.biomes)]);

			std::vector<uint16_t> palette_biomes(paletteb.payload.size());
			size_t i = 0;
			for (auto pbit = paletteb.payload.begin(); pbit != paletteb.payload.end(); ++pbit, ++i) {
				nbt::TagString& biome = (*pbit)->cast<nbt::TagString>();
				palette_biomes[i] = mapcrafter::renderer::Biome::getBiomeId(biome.payload);
			}
			// Convert chunk local index into the global biome index
			for (i = 0; i < boost::size(section.biomes); ++i) {
				uint16_t idx = section.biomes[i];
				// Make sure we stay in the array, if it happens, use the default biome
				if (idx>=palette_biomes.size()) idx = 0;
				section.biomes[i] = palette_biomes[idx];
			}
		} else if (paletteb.payload.size()==1) {
			// Only 1 in palette: It's only this biome in this chunk
			nbt::TagString& biome = (*paletteb.payload.begin())->cast<nbt::TagString>();
			std::fill(section.biomes, section.biomes+boost::size(section.biomes), mapcrafter::renderer::Biome::getBiomeId(biome.payload));
		} else {
			// No palette, this shouldn't happen, anyway let's use the default one
			std::fill(section.biomes, section.biomes+boost::size(section.biomes), 0);
		}

		if (section_tag.hasArray<nbt::TagByteArray>("BlockLight")) {
			const nbt::TagByteArray& block_light = section_tag.findTag<nbt::TagByteArray>("BlockLight");
			std::copy(block_light.payload.begin(), block_light.payload.end(), section.block_light);
		} else {
			std::fill(&section.block_light[0], &section.block_light[2048], 0);
		}

		if (section_tag.hasArray<nbt::TagByteArray>("SkyLight", 2048)) {
			const nbt::TagByteArray& sky_light = section_tag.findTag<nbt::TagByteArray>("SkyLight");
			std::copy(sky_light.payload.begin(), sky_light.payload.end(), section.sky_light);
		} else {
			std::fill(&section.sky_light[0], &section.sky_light[2048], 0);
		}

		// add this section to the section list
		section_offsets[section.y-CHUNK_LOWEST] = sections.size();
		sections.push_back(section);
	}

	return true;
}

void Chunk::clear() {
	sections.clear();
	for (size_t i = 0; i < boost::size(section_offsets); i++)
		section_offsets[i] = -1;
}

bool Chunk::hasSection(int y) const {
	const ChunkSection* cs = this->getSection(y);
	return cs != NULL;
}

const ChunkSection* Chunk::getSection(int y) const {
	int chunk_idx = y >> 4;
	if( chunk_idx < CHUNK_LOWEST || chunk_idx >= CHUNK_HIGHEST) {
		return NULL;
	}
	// Convert index to index into the data array
	size_t section_idx = section_offsets[chunk_idx - CHUNK_LOWEST];
	if( section_idx < 0 || section_idx >= sections.size()) {
		return NULL;
	}
	return &sections[section_idx];
}

uint16_t Chunk::getBlockID(const LocalBlockPos& pos, bool force) const {
	const ChunkSection* cs = getSection(pos.y);
	if (!cs)
		return nop_id;

	int x = pos.x;
	int z = pos.z;

	// check whether this block is really rendered
	if (!checkBlockWorldCrop(x, z, pos.y))
		return nop_id;

	// calculate the offset and get the block ID
	// and don't forget the add data
	int offset = ((pos.y & 15) * 256) + (z * 16) + x;
	uint16_t id = cs->block_ids[offset];
	if (!force && world_crop.hasBlockMask()) {
		const BlockMask* mask = world_crop.getBlockMask();
		BlockMask::BlockState block_state = mask->getBlockState(id);
		if (block_state == BlockMask::BlockState::COMPLETELY_HIDDEN)
			return nop_id;
		else if (block_state == BlockMask::BlockState::COMPLETELY_SHOWN)
			return id;
		if (mask->isHidden(id, 0 /*getBlockData(pos, true)*/))
			return nop_id;
	}
	return id;
}

bool Chunk::checkBlockWorldCrop(int x, int z, int y) const {
	// now about the actual world cropping:
	// get the global position of the block
	BlockPos global_pos = LocalBlockPos(x, z, y).toGlobalPos(chunkpos);
	// check whether the block is contained in the y-bounds.
	if (!world_crop.isBlockContainedY(global_pos))
		return false;
	// only check x/z-bounds if the chunk is not completely contained
	if (!chunk_completely_contained && !world_crop.isBlockContainedXZ(global_pos))
		return false;
	return true;
}

uint8_t Chunk::getData(const LocalBlockPos& pos, int array, bool force) const {
	const ChunkSection* cs = getSection(pos.y);
	if (!cs) {
		 // not existing sections should always have skylight
		 return array == 1 ? 15 : 0;
	}

	int x = pos.x;
	int z = pos.z;

	// check whether this block is really rendered
	if (!checkBlockWorldCrop(x, z, pos.y)) {
		 return array == 1 ? 15 : 0;
	}

	uint8_t data = 0;
	// calculate the offset and get the block data
	int offset = ((pos.y & 15) * 256) + (z * 16) + x;
	// handle bottom/top nibble
	if ((offset % 2) == 0)
		 data = cs->getArray(array)[offset / 2] & 0xf;
	else
		 data = (cs->getArray(array)[offset / 2] >> 4) & 0x0f;
	if (!force && world_crop.hasBlockMask()) {
		 const BlockMask* mask = world_crop.getBlockMask();
		 if (mask->isHidden(getBlockID(pos, true), data)) {
			  return array == 1 ? 15 : 0;
		}
	}
	return data;
}

uint8_t Chunk::getBlockLight(const LocalBlockPos& pos) const {
	return getData(pos, 0);
}

uint8_t Chunk::getSkyLight(const LocalBlockPos& pos) const {
	return getData(pos, 1);
}

uint16_t Chunk::getBiomeAt(const LocalBlockPos& pos) const {
	const ChunkSection* cs = getSection(pos.y);
	if (!cs)
		return renderer::DEFAULT_BIOME_ID;

	int x = pos.x >> 2;
	int z = pos.z >> 2;
	int y = (pos.y & 15) >> 2;

	return cs->biomes[(y << 4) + (z << 2) + x];
}

const ChunkPos& Chunk::getPos() const {
	return chunkpos;
}

}
}
